siva/storage.go

```diff
@@ -397,5 +412,77 @@ func (s *Storage) PackRefs() (err error) {
 		}
 	}
 
-	return butil.RemoveAll(s.fs, refsPath)
+	return s.removeRefsDir()
+}
+
+func (s *Storage) removeRefsDir() error {
+	fs := s.Filesystem()
+	_, err := fs.Stat(refsPath)
+	if err != nil {
+		if fs == s.transFS {
+			fs = s.baseFS
+			_, err = fs.Stat(refsPath)
+
+		}
+
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+
+			if err := fs.MkdirAll(refsPath, 0770); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := createKeepFile(fs, refsPath); err != nil {
+		return err
+	}
+
+	entries, err := fs.ReadDir(refsPath)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		if e.Name() == keepFile && e.Mode().IsRegular() {
+			continue
+		}
+
+		if err := butil.RemoveAll(
+			fs,
+			filepath.Join(refsPath, e.Name()),
+		); err != nil {
+			return err
+		}
+	}
+
+	if s.transactional {
+		if err := s.baseFS.Sync(); err != nil {
```

Comment:
Sync should only be done if needed. That is, if there are no previous unpacked references and `refs` dir already exists. If I understand correctly this code is always generating to indexes per transaction. One with the delete unpacked refs + create keep file and another with the proper transaction.